src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -861,5 +861,50 @@ unsafe void M()
                     //         delegate*<C.D> d;
                     Diagnostic(ErrorCode.WRN_UnreferencedVar, "d").WithArguments("d").WithLocation(10, 24));
         }
+
+        [Fact]
+        public void FunctionPointerConstraintIntroducedBySubstitution()
+        {
+            string source = @"
+class R1<T1>
+{
+    public virtual void f<T2>() where T2 : T1 { }
+}
+class R2 : R1<delegate*<void>>
+{
+    public override void f<T2>() { }
+}
+class Program
+{
+    static void Main(string[] args)
+    {
+        R2 r = new R2();
+        r.f<int>();
+    }
+}";
+
+            var compilation = CreateFunctionPointerCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (6,7): error CS0306: The type 'delegate*<void>' may not be used as a type argument
+                // class R2 : R1<delegate*<void>>
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "R2").WithArguments("delegate*<void>").WithLocation(6, 7)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeAsThisOfExtensionMethod()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe static class C
+{
+    static void M(this delegate*<void> ptr) {}
```

Comment:
>static void M(this delegate*<void> ptr) {} [](start = 4, length = 42)

Is it worth trying to define a method like this in metadata and trying to consume it?

Reply:
Probably, I'll do that.

---
In reply to: [407755830](https://github.com/dotnet/roslyn/pull/43248#discussion_r407755830) [](ancestors = 407755830)

Reply:
To see what we do for regular pointers, I created IL that defined an extension method on `int*`. Turns out, we just allow it with no errors, only defining such a method in C# is illegal.

---
In reply to: [407803051](https://github.com/dotnet/roslyn/pull/43248#discussion_r407803051) [](ancestors = 407803051,407755830)

Reply:
> Turns out, we just allow it with no errors, only defining such a method in C# is illegal.

I am curious what would be the motivation for the error if we are able to consume such methods just fine.

---
In reply to: [407808055](https://github.com/dotnet/roslyn/pull/43248#discussion_r407808055) [](ancestors = 407808055,407803051,407755830)